### PR TITLE
Add configurable parameters for all the required fields

### LINKED ISSUE #522: KubeEdge Installer: Add configurable parameters for all the required fields in controller.yaml and edge.yaml
**What would you like to be added**:
Adding configurable parameters to KubeEdge init and join, so that it can be used to populate the required fields in controller.yaml and edge.yaml.

**Why is this needed**:
It gives more and better user experience to run own efficient KubeEdge environment and deployment.

### PR DESCRIPTION
**Pull request by kadisi**
Signed-off-by: zhangjie <iamkadisi@163.com>

**What type of PR is this?**

/kind feature
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #522 

**Special notes for your reviewer**:

### CODE REVIEW

File: keadm/app/cmd/edge/join.go [hunk 1]
```diff
@@ -102,14 +106,47 @@ func addJoinOtherFlags(cmd *cobra.Command, joinOptions *types.JoinOptions) {
...
+	cmd.Flags().Uint16Var(&joinOptions.WebSocketPort, types.WebSocketPort, joinOptions.WebSocketPort,
+		"WebSocket port of kubeEdge edgecontroller, default: 10000")
+
+	cmd.Flags().Uint16Var(&joinOptions.QuicPort, types.QuicPort, joinOptions.QuicPort,
+		"Quic port of kubeEdge edgecontroller, default: 10001")
+
+	cmd.Flags().StringVar(&joinOptions.EdgeYaml.EdgeHub.Controller.Protocal, types.Protocal, joinOptions.EdgeYaml.EdgeHub.Controller.Protocal,
+		"Protocal between edge_core and edgecontroller, support websocket and quic , default: websocket")
+
+	joinOptions.EdgeYaml.EdgeHub.WebSocket.URL = fmt.Sprintf("wss://%s:10000/e632aba927ea4ac2b575ec1603d56f10/%s/events",
```
Comment (samy2019) [inline keadm/app/cmd/edge/join.go]: This string is used in multiple places, Hence define a constant string and used it in different places.

File: keadm/app/cmd/edge/join.go [hunk 2]
```diff
@@ -102,14 +106,47 @@ func addJoinOtherFlags(cmd *cobra.Command, joinOptions *types.JoinOptions) {
...
-		KubernetesVersion: types.DefaultK8SVersion}
-	opts.CertPath = types.DefaultCertPath
+	opts := &types.JoinOptions{
+		GlobalOptions: types.GlobalOptions{
+			DockerVersion:     types.DefaultDockerVersion,
+			KubeEdgeVersion:   types.DefaultKubeEdgeVersion,
+			KubernetesVersion: types.DefaultK8SVersion},
+		EdgeYaml: types.NewEdgeYamlSt(),
+		LogYaml: &types.LoggingYaml{
+			LoggerLevel:   "DEBUG",
```
Comment (samy2019) [inline keadm/app/cmd/edge/join.go]: Chang the log level to "INFO"

**Comment by sids-b** [comment]
ping @kadisi for addressing the review comments

File: keadm/app/cmd/common/types.go [hunk 3]
```diff
@@ -153,6 +168,42 @@ type ControllerYaml struct {
...
+				Namespace:           constants.DefaultKubeNamespace,
+				ContentType:         constants.DefaultKubeContentType,
+				QPS:                 constants.DefaultKubeQPS,
+				Burst:               constants.DefaultKubeBurst,
+				NodeUpdateFrequency: constants.DefaultKubeUpdateNodeFrequency * time.Second,
+			}},
+		CloudHub: CloudHubSt{
+			IPAddress:         "0.0.0.0",
+			Port:              10000,
+			QuicPort:          100001,
```
[the remark this answers is not in the document]
Comment (kadisi) [reply keadm/app/cmd/common/types.go]: ok

File: keadm/app/cmd/common/constant.go [hunk 4]
```diff
@@ -26,15 +26,33 @@ const (
...
+	KubeMaster = "controller.kube.master"
 
 	//K8SAPIServerIPPort sets the IP:Port of Kubernetes api-server
 	K8SAPIServerIPPort = "k8sserverip"
 
 	//EdgeControllerIP sets the IP of KubeEdge edgecontroller cloud component
 	EdgeControllerIP = "edgecontrollerip"
 
+	//Protocal set the connection protocal between edge and edgecontroller
+	Protocal = "protocal"
```
[the remark this answers is not in the document]
Comment (kadisi) [reply keadm/app/cmd/common/constant.go]: Protocol = "protocol"  ?

File: keadm/app/cmd/common/types.go [hunk 5]
```diff
@@ -153,6 +168,42 @@ type ControllerYaml struct {
...
+			CA:                "/etc/kubeedge/ca/rootCA.crt",
+			Cert:              "/etc/kubeedge/certs/edge.crt",
+			Key:               "/etc/kubeedge/certs/edge.key",
+			KeepAliveInterval: 30,
+			WriteTimeout:      30,
+			NodeLimit:         10,
```
[the remark this answers is not in the document]
Comment (kadisi) [reply keadm/app/cmd/common/types.go]: ok

File: keadm/app/cmd/common/types.go [hunk 6]
```diff
@@ -250,3 +307,51 @@ type EdgeYamlSt struct {
...
+	edgeID := "fb4ebb70-2783-42b8-b3ef-63e2fd6d242e"
+	url := "wss://0.0.0.0:10000/e632aba927ea4ac2b575ec1603d56f10/fb4ebb70-2783-42b8-b3ef-63e2fd6d242e/events"
+	version := "2.0.0"
+
+	return &EdgeYamlSt{
+		MQTT: MQTTConfig{
+			Server:           "tcp://127.0.0.1:1883",
+			InternalServer:   "tcp://127.0.0.1:1884",
+			Mode:             MQTTInternalMode,
+			QOS:              MQTTQoSAtMostOnce,
```
[the remark this answers is not in the document]
Comment (kadisi) [reply keadm/app/cmd/common/types.go]: ok

File: keadm/app/cmd/edge/join.go [hunk 7]
```diff
@@ -102,14 +106,47 @@ func addJoinOtherFlags(cmd *cobra.Command, joinOptions *types.JoinOptions) {
...
+	cmd.Flags().StringVar(&joinOptions.EdgeYaml.EdgeHub.Controller.Protocal, types.Protocal, joinOptions.EdgeYaml.EdgeHub.Controller.Protocal,
+		"Protocal between edge_core and edgecontroller, support websocket and quic , default: websocket")
+
+	joinOptions.EdgeYaml.EdgeHub.WebSocket.URL = fmt.Sprintf("wss://%s:10000/e632aba927ea4ac2b575ec1603d56f10/%s/events",
+		joinOptions.EdgeControllerIP, joinOptions.EdgeNodeID)
+	joinOptions.EdgeYaml.EdgeHub.Controller.NodeID = joinOptions.EdgeNodeID
+	joinOptions.EdgeYaml.EdgeD.HostnameOverride = joinOptions.EdgeNodeID
+
+	joinOptions.EdgeYaml.EdgeHub.Quic.URL = fmt.Sprintf("%s:%s", joinOptions.EdgeControllerIP, joinOptions.QuicPort)
+
```
[the remark this answers is not in the document]
Comment (kadisi) [reply keadm/app/cmd/edge/join.go]: ok

File: keadm/app/cmd/edge/join.go (hunk 1 above)
Comment (kadisi) [reply keadm/app/cmd/edge/join.go]: ok

File: keadm/app/cmd/common/types.go [hunk 8]
```diff
@@ -212,8 +261,13 @@ type WebSocketSt struct {
 	ReadDeadline     uint16 `yaml:"read-deadline"`
 }
 
+type QuicSt struct {
+	URL string `yaml:"url"`
+}
+
 //ControllerSt contain edgecontroller config which edge component uses
 type ControllerSt struct {
+	Protocal            string `yaml:"protocal"`
```
Comment (samy2019) [inline keadm/app/cmd/common/types.go]: Change Protocal to Protocol. 
Change protocal to protocol.

File: keadm/app/cmd/common/types.go [hunk 9]
```diff
@@ -250,3 +305,56 @@ type EdgeYamlSt struct {
...
+
+func NewEdgeYamlSt() *EdgeYamlSt {
+	edgeID := "fb4ebb70-2783-42b8-b3ef-63e2fd6d242e"
+	url := CreateWebSocketUrl("0.0.0.0", uint16(DefaultWebSocketPort), "fb4ebb70-2783-42b8-b3ef-63e2fd6d242e/")
+	version := "2.0.0"
+
+	return &EdgeYamlSt{
+		MQTT: MQTTConfig{
+			Server:           "tcp://127.0.0.1:1883",
+			InternalServer:   "tcp://127.0.0.1:1884",
```
Comment (samy2019) [inline keadm/app/cmd/common/types.go]: Replace these with the newly defined constants @ keadm/app/cmd/common/constant.go

File: keadm/app/cmd/common/types.go [hunk 10]
```diff
@@ -250,3 +305,56 @@ type EdgeYamlSt struct {
...
+			Retain:           false,
+			SessionQueueSize: 100},
+		EdgeHub: EdgeHubSt{
+			WebSocket: WebSocketSt{
+				URL:              url,
+				CertFile:         "/etc/kubeedge/certs/edge.crt",
+				KeyFile:          "/etc/kubeedge/certs/edge.key",
+				HandshakeTimeout: 30,
+				WriteDeadline:    15,
+				ReadDeadline:     15},
```
Comment (samy2019) [inline keadm/app/cmd/common/types.go]: Replace these with the newly defined constants @ keadm/app/cmd/common/constant.go

File: keadm/app/cmd/edge/join.go [hunk 11]
```diff
@@ -102,14 +106,55 @@ func addJoinOtherFlags(cmd *cobra.Command, joinOptions *types.JoinOptions) {
...
+			LogFormatText: true,
+			Writers:       []string{"stdout"}},
+		ModulesYaml: &types.ModulesYaml{
+			Modules: types.ModulesSt{
+				Enabled: []string{"eventbus", "servicebus", "websocket", "metaManager", "edged", "twin"}},
+		},
+		//If the user doesn't provide any edge ID on the command line, then it generates unique id and assigns it.
+		EdgeNodeID:    uuid.New().String(),
+		WebSocketPort: 10000,
+		QuicPort:      10001,
```
Comment (samy2019) [inline keadm/app/cmd/edge/join.go]: Replace these with the newly defined constants @ keadm/app/cmd/common/constant.go

File: keadm/app/cmd/edge/join.go [hunk 12]
```diff
@@ -87,6 +89,8 @@ func addJoinOtherFlags(cmd *cobra.Command, joinOptions *types.JoinOptions) {
 
 	cmd.Flags().StringVar(&joinOptions.KubeEdgeVersion, types.KubeEdgeVersion, joinOptions.KubeEdgeVersion,
 		"Use this key to download and use the required KubeEdge version")
+	joinOptions.EdgeYaml.EdgeD.Version = joinOptions.KubernetesVersion
```
Comment (samy2019) [inline keadm/app/cmd/edge/join.go]: This should be joinOptions.KubeEdgeVersion and not joinOptions.KubernetesVersion. Please fix it.

File: keadm/app/cmd/edge/join.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (samy2019) [reply keadm/app/cmd/edge/join.go]: Is this comment addressed. I can't see any flag for log level setting. 
cmd.Flags().StringVar(&initOpts.LogYaml.LoggerLevel, types.LoggerLevel, initOpts.LogYaml.LoggerLevel,
		"Use this key to set loggerlevel, default: INFO")

File: keadm/app/cmd/cloud/init.go [hunk 13]
```diff
@@ -101,8 +114,24 @@ func addJoinOtherFlags(cmd *cobra.Command, initOpts *types.InitOptions) {
...
+		"Use this key to set controller.kube.kubeconfig path, eg: $HOME/.kube/config")
+
+	cmd.Flags().StringVar(&initOpts.ControllerYaml.DeviceController.Kube.Master, types.DeviceControllerKubeConfig, initOpts.ControllerYaml.DeviceController.Kube.KubeConfig,
+		"Use this key to set devicecontroller.kube.kubeconfig path, eg: $HOME/.kube/config")
+
+	cmd.Flags().Uint16Var(&initOpts.ControllerYaml.CloudHub.Port, types.WebSocketPort, initOpts.ControllerYaml.CloudHub.Port,
+		"Use this key to set cloudhub websocket port, default: 10000")
+
+	cmd.Flags().Uint16Var(&initOpts.ControllerYaml.CloudHub.QuicPort, types.QuicPort, initOpts.ControllerYaml.CloudHub.QuicPort,
+		"Use this key to set cloudhub quict port, default: 10001")
```
Comment (rohitsardesai83) [inline keadm/app/cmd/cloud/init.go]: quict  -> quic

File: keadm/app/cmd/cloud/init.go [hunk 14]
```diff
@@ -101,8 +114,24 @@ func addJoinOtherFlags(cmd *cobra.Command, initOpts *types.InitOptions) {
...
 
-	cmd.Flags().StringVar(&initOpts.KubeConfig, types.KubeConfig, initOpts.KubeConfig,
-		"Use this key to set kube-config path, eg: $HOME/.kube/config")
+	cmd.Flags().StringVar(&initOpts.ControllerYaml.Controller.Kube.Master, types.KubeMaster, initOpts.ControllerYaml.Controller.Kube.Master,
+		"Use this key to set controller.kube.master , eg: http://localhost:8080")
+
+	cmd.Flags().StringVar(&initOpts.ControllerYaml.Controller.Kube.KubeConfig, types.ControllerKubeKubeconfig, initOpts.ControllerYaml.Controller.Kube.KubeConfig,
+		"Use this key to set controller.kube.kubeconfig path, eg: $HOME/.kube/config")
+
+	cmd.Flags().StringVar(&initOpts.ControllerYaml.DeviceController.Kube.Master, types.DeviceControllerKubeConfig, initOpts.ControllerYaml.DeviceController.Kube.KubeConfig,
```
Comment (rohitsardesai83) [inline keadm/app/cmd/cloud/init.go]: I think we can optimize by making kube config as a common configuration in the yaml .  I am not sure if we have an open issue for it.

File: keadm/app/cmd/common/constant.go [hunk 15]
```diff
@@ -52,4 +82,43 @@ const (
 
 	//DefaultK8SVersion is the current default version of K8S
 	DefaultK8SVersion = "1.14.1"
+
+	// DefaultWebSocketPort is default websocket port
+	DefaultWebSocketPort = 10000
+
+	// DefaultQuitPort is default quic port
+	DefaultQuitPort = 10001
```
Comment (rohitsardesai83) [inline keadm/app/cmd/common/constant.go]: Quit -> Quic

File: keadm/app/cmd/common/types.go [hunk 16]
```diff
@@ -17,24 +17,39 @@ limitations under the License.
...
+	ModulesYaml    *ModulesYaml
 }
 
 //JoinOptions has the kubeedge cloud init information filled by CLI
 type JoinOptions struct {
-	InitOptions
-	CertPath           string
-	EdgeControllerIP   string
+	GlobalOptions
+	EdgeYaml    *EdgeYamlSt
```
Comment (rohitsardesai83) [inline keadm/app/cmd/common/types.go]: Can we rename all structures ending with "St" to something more meaningful ?
Such suffixes are not easy to perceive what they stand for .

File: keadm/app/cmd/common/types.go [hunk 17]
```diff
@@ -250,3 +305,56 @@ type EdgeYamlSt struct {
 	EdgeHub EdgeHubSt  `yaml:"edgehub"`
 	EdgeD   EdgeDSt    `yaml:"edged"`
 }
+
+// CreateWebSocketUrl create websocket url
+func CreateWebSocketUrl(address string, port uint16, id string) string {
+	return fmt.Sprintf("wss://%s:%v/e632aba927ea4ac2b575ec1603d56f10/%s/events", address, port, id)
```
Comment (rohitsardesai83) [inline keadm/app/cmd/common/types.go]: e632aba927ea4ac2b575ec1603d56f10  --> can this be configured as well ?

File: keadm/app/cmd/common/types.go [hunk 18]
```diff
@@ -250,3 +305,56 @@ type EdgeYamlSt struct {
...
+}
+
+func NewEdgeYamlSt() *EdgeYamlSt {
+	edgeID := "fb4ebb70-2783-42b8-b3ef-63e2fd6d242e"
+	url := CreateWebSocketUrl("0.0.0.0", uint16(DefaultWebSocketPort), "fb4ebb70-2783-42b8-b3ef-63e2fd6d242e/")
+	version := "2.0.0"
```
Comment (rohitsardesai83) [inline keadm/app/cmd/common/types.go]: we need to get the version from the release package ? I am not which branch or release does "2.0.0" represent.

File: keadm/app/cmd/common/types.go [hunk 19]
```diff
@@ -250,3 +305,56 @@ type EdgeYamlSt struct {
...
+				ReadDeadline:     DefaultReadDeadline},
+			Quic: QuicSt{
+				URL: "127.0.0.1:10001",
+			},
+			Controller: ControllerSt{
+				Protocol:            "websocket",
+				Placement:           false,
+				Heartbeat:           15,
+				RefreshAKSKInterval: 10,
+				AuthInfoFilesPath:   "/var/IEF/secret",
```
Comment (rohitsardesai83) [inline keadm/app/cmd/common/types.go]: please investigate if these can be kept optional. These are Huawei cloud specific .

File: keadm/app/cmd/common/types.go [hunk 20]
```diff
@@ -250,3 +305,56 @@ type EdgeYamlSt struct {
...
+				PlacementURL:        "https://10.154.193.32:7444/v1/placement_external/message_queue",
```
Comment (rohitsardesai83) [inline keadm/app/cmd/common/types.go]: same comment as above.

File: keadm/app/cmd/cloud/init.go (hunk 14 above)
Comment (kadisi) [reply keadm/app/cmd/cloud/init.go]: @rohitsardesai83  i agree with you.    

@sids-b @samy2019  what do you think?

File: keadm/app/cmd/common/constant.go (hunk 15 above)
Comment (kadisi) [reply keadm/app/cmd/common/constant.go]: ok

File: keadm/app/cmd/common/types.go (hunk 16 above)
Comment (kadisi) [reply keadm/app/cmd/common/types.go]: ok

File: keadm/app/cmd/common/types.go (hunk 17 above)
Comment (kadisi) [reply keadm/app/cmd/common/types.go]: ok

File: keadm/app/cmd/common/types.go (hunk 18 above)
Comment (kadisi) [reply keadm/app/cmd/common/types.go]: @samy2019 need your help.

File: keadm/app/cmd/common/types.go (hunk 20 above)
Comment (kadisi) [reply keadm/app/cmd/common/types.go]: ok

**Comment by sids-b** [comment]
@kadisi , can you please resolve the conflicts and address the review comments ?

**Comment by sids-b** [comment]
@kadisi , please resolve the conflicts and resolve the comments that are already addressed. The PR looks good to me.

**Comment by kadisi** [comment]
@sids-b PTAL

**Comment by rohitsardesai83** [comment]
@kadisi , have all comments been addressed ? Can you resolve the conversations ?

**Pull request closed without merging**